README.md
# Questionnaire

Package for generating random question(s) in multiple languages.

[![Testing](https://github.com/MilosPaunovic/questionnaire/actions/workflows/testing.yml/badge.svg)](https://github.com/MilosPaunovic/questionnaire/actions/workflows/testing.yml)

## Try it Online

[Link to Codepen playground](https://codepen.io/milospaunovic/pen/MWGrdEZ?editors=1011)

## Motivation

Main idea and motivation behind this was work from [Nenad Ilić](https://github.com/ilic5000) and his post I saw on [Reddit](https://www.reddit.com/r/programiranje/comments/xofp1t/anansi_computer_vision_mini_projekat_izvukao_sam/). You could (and should) read more about it [here](https://github.com/ilic5000/pabkvizgenerator/blob/main/README.md).

## Installation

```bash
npm i @paunovic/questionnaire
```

## Configuring

```js
import { initialize } from '@paunovic/questionnaire';

// Initialization with default language pack
const QUESTIONNAIRE = initialize();

// Initialization with different language pack, including variation
const QUESTIONNAIRE = initialize({ countryCode: 'rs', variation: 'cyrillic' });
```

Currently available country codes and variations:

| Country       | Country code | Variation  |  Questions  |
| :------------ | :----------: | :--------: | :---------: |
| Serbia `*`    | `rs`         | `cyrillic` | 10253       |

`*` Default language; if no `countryCode` parameter is passed to `initialize` method, Serbian Latin will be loaded as a default language.

## Methods

#### Question

Returns random question from the list in the object form, containing question and answer properties.

```js
QUESTIONNAIRE.question(); // => { question: 'QUESTION', answer: 'ANSWER' }
```

Returns random question from the list in the object form, containing question, answer and wrong (containg number of wrong answers you've specified, in range from 1 to 30) properties.

```js
QUESTIONNAIRE.question({ wrong: 3}); // => { question: 'QUESTION', answer: 'ANSWER', wrong: ['WRONG', 'WRONG', 'WRONG'] }
```


#### Questions

Returns an array of 10 question objects from the list.

```js
QUESTIONNAIRE.questions() // => [{ question: 'QUESTION', answer: 'ANSWER' }, { question: 'QUESTION', answer: 'ANSWER' }...]
```

Returns an array of question objects with wrong answers (containg number of wrong answers you've specified, in range from 1 to 30) from the list (number of items defaults to 10, accepted range from 1 to 10000).

```js
QUESTIONNAIRE.questions({ howMany: 7, wrong: 1 }) // => [{ question: 'QUESTION', answer: 'ANSWER', wrong: ['WRONG'] }, { question: 'QUESTION', answer: 'ANSWER', wrong: ['WRONG'] }...]
```

## Contributing

Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change. Please try to follow [semantic commit messages](https://gist.github.com/joshbuchea/6f47e86d2510bce28f8e7f42ae84c716) principle to simplify navigation through git history.

## License

[MIT](https://raw.githubusercontent.com/MilosPaunovic/questionnaire/develop/LICENSE) © [Milos Paunovic](https://github.com/MilosPaunovic)


CHANGELOG.md
# Changelog

## [3.0.0](https://github.com/MilosPaunovic/questionnaire/compare/2.0.0...3.0.0) - 2022-09-30

### Breaking changes

Dropped Node v12 support.

## [2.0.0](https://github.com/MilosPaunovic/questionnaire/compare/1.0.4...2.0.0) - 2022-09-30

### Breaking changes

This major release introduces a breaking change as the method `questions()` now expects an object parameter, instead of number, like in the previous versions.

### Added

* Allowed options argument to be passed for recieving wrong answers, as well as the correct one. 

## [1.0.4](https://github.com/MilosPaunovic/questionnaire/compare/1.0.3...1.0.4) - 2022-09-28

### Fixed

* Included question data files into the final bundle.

## [1.0.3](https://github.com/MilosPaunovic/questionnaire/compare/1.0.2...1.0.3) - 2022-09-28

### Changed

* Improved the way of library bundling.

### Removed

* Trimmed down list of questions following the [advice](https://www.linkedin.com/feed/update/urn:li:activity:6980753295447330816?commentUrn=urn%3Ali%3Acomment%3A%28activity%3A6980753295447330816%2C6980847167380803586%29) of the initial answers creator.

## [1.0.2](https://github.com/MilosPaunovic/questionnaire/commits/1.0.2) - 2022-09-28

_Initial release._


test/lib.js
import { assert } from 'chai';
import { initialize } from '../lib/index.js';

const QUESTIONNAIRE = initialize()

describe('Testing basic functionalities of the entire library', () => {

	it('question()  - Should return a QUESTIONNAIRE question', () => {
		const result = QUESTIONNAIRE.question();
		assert.isObject(result) && assert.isNotEmpty(result);
	});

	it('questions() - Should return array of 10 random questions (no parameter passed)', () => {
		const result = QUESTIONNAIRE.questions();
		assert.isArray(result) && assert.lengthOf(result, 10);
	});

	it('questions() - Should return array of 92 random questions (with parameter passed)', () => {
		const result = QUESTIONNAIRE.questions(92);
		assert.isArray(result) && assert.lengthOf(result, 92);
	});

	it('questions() - Should throw an error (invalid parameter passed)', () => {
		assert.throw(() => { QUESTIONNAIRE.questions("invalid"); }, Error, 'Parameter howMany must be convertible to number.');
	});

});


test/general.js
import { describe, it } from 'mocha';
import { expect, assert } from 'chai';
import { initialize } from '../src/index.js';

describe('Testing basic functionalities of the entire library', () => {
	it('Should successfully initialize with default parameters', () => {
		const QUESTIONNAIRE = initialize();
		expect(Object.keys(QUESTIONNAIRE)).not.to.be.empty;
	});

	it('Should successfully initialize with parameters passed', () => {
		const QUESTIONNAIRE = initialize({ countryCode: 'rs', variation: 'cyrillic' });
		expect(Object.keys(QUESTIONNAIRE)).not.to.be.empty;
	});

	it('Should throw an error (invalid countryCode parameter passed)', () => {
		assert.throw(() => { initialize({ countryCode: 'INVALID' }); }, Error, 'Parameter countryCode is not valid.');
	});

	it('Should throw an error (invalid variation parameter passed, without countryCode)', () => {
		assert.throw(() => { initialize({ variation: 'INVALID' }); }, Error, 'Parameter variation is not valid.');
	});

	it('Should throw an error (invalid variation parameter passed, with countryCode)', () => {
		assert.throw(() => { initialize({ countryCode: 'rs', variation: 'INVALID' }); }, Error, 'Parameter variation is not valid.');
	});
});


test/question.js
import { describe, it } from 'mocha';
import { assert, expect } from 'chai';
import { initialize } from '../src/index.js';

const QUESTIONNAIRE = initialize();

describe('Testing question() method', () => {
	it('Return a random question', () => {
		const result = QUESTIONNAIRE.question();
		assert.isObject(result) && assert.isNotEmpty(result);
	});

	it('Ignore parameter type different than object', () => {
		const result = QUESTIONNAIRE.question('INVALID');
		assert.isObject(result) && assert.isNotEmpty(result);
	});

	it('Ignore empty object parameter', () => {
		const result = QUESTIONNAIRE.question({});
		assert.isObject(result) && assert.isNotEmpty(result);
	});

	it('Ignore object parameter with invalid key/value pair', () => {
		const result = QUESTIONNAIRE.question({ INVALID: 'INVALID' });
		assert.isObject(result) && assert.isNotEmpty(result);
	});

	it('Invalid number of wrong answers', () => {
		assert.throw(() => { QUESTIONNAIRE.question({ wrong: 'INVALID' }); }, Error, 'Parameter wrong must be castable to integer.');
	});

	it('Valid number of wrong answers', () => {
		const result = QUESTIONNAIRE.question({ wrong: 5 });
		assert.isObject(result) && assert.isNotEmpty(result) && expect(result).to.have.key('wrong') && expect(result.wrong).to.have.lengthOf(5);
	});

	it('Smaller number of wrong answers, should default to 1', () => {
		const result = QUESTIONNAIRE.question({ wrong: -1 });
		assert.isObject(result) && assert.isNotEmpty(result) && expect(result).to.have.key('wrong') && expect(result.wrong).to.have.lengthOf(1);
	});

	it('Larger number of wrong answers, should default to 30', () => {
		const result = QUESTIONNAIRE.question({ wrong: 31 });
		assert.isObject(result) && assert.isNotEmpty(result) && expect(result).to.have.key('wrong') && expect(result.wrong).to.have.lengthOf(30);
	});
});


src/helpers/general.js
/**
 * Returns random number in range between 1 and supplied maximum.
 *
 * @param {number} maximum Upper limit of generated random number.
 * @return {number} Random number between 1 and suplied maximum.
 */
export const randomNumber = (maximum) => {
	if (!maximum) throw new Error('Parameter maximum is missing.');
	if (!parseInt(maximum, 10)) throw new Error('Parameter maximum must be castable to integer.');

	return Math.floor(Math.random() * parseInt(maximum, 10)) + 1;
};

/**
 * Returns an array of random numbers in range between 1 and supplied maximum.
 *
 * @param {number} items Number of array items to be returned.
 * @param {number} maximum Upper limit of generated random number.
 * @return {number} Array of random numbers between 1 and suplied maximum.
 */
export const randomNumbers = (items, maximum) => {
	if (!items) throw new Error('Parameter items is missing.');
	if (!parseInt(items, 10)) throw new Error('Parameter items must be castable to integer.');

	if (!maximum) throw new Error('Parameter maximum is missing.');
	if (!parseInt(maximum, 10)) throw new Error('Parameter maximum must be castable to integer.');

	const result = new Set();

	while (result.size < items) {
		result.add(randomNumber(maximum));
	}

	return Array.from(result);
};

/**
 * Checks if the parameter can be casted to integer and, if so, limits the
 * value to supplied minimum and maximum.
 *
 * @param {any} value Parameter value itself.
 * @param {string} name Name of the parameter for error message.
 * @param {number} minimum Lower limit of value.
 * @param {number} maximum Upper limit of value.
 * @return {number} Properly casted and limited value.
 */
export const parameterLimitter = (value, name, minimum, maximum) => {
	if (!parseInt(value, 10)) throw new Error(`Parameter ${name} must be castable to integer.`);

	if (parseInt(value, 10) < minimum) value = minimum;
	if (parseInt(value, 10) > maximum) value = maximum;

	return value;
};


src/index.js
/*
 * @paunovic/questionnaire@2.0.0
 * (c) Milos Paunovic
 * Released under the MIT License.
 */

import { LANGUAGES } from './helpers/languages.js';
import { randomNumber, randomNumbers, parameterLimitter } from './helpers/general.js';

/**
 * Initialize the package by selecting language and variation to use.
 *
 * @param {object} details Details for library initialization.
 * @param {string} [details.countryCode=rs] 2 letter ISO country code.
 * @param {string} [details.variation=default] Variation of the language.
 * @return {object} Available functions to be consumed.
 */
export const initialize = ({ countryCode = 'rs', variation = 'default' } = {}) => {
	if (!LANGUAGES[countryCode]) throw new Error('Parameter countryCode is not valid.');
	else if (!LANGUAGES[countryCode][variation]) throw new Error('Parameter variation is not valid.');

	const QUESTIONS = LANGUAGES[countryCode][variation];
	const AVAILABLE = QUESTIONS.length - 1;

	/**
	 * Returns random question object from the list.
	 *
	 * @param {object} [options] Guidelines for question generation.
	 * @param {number} [options.wrong] Number of wrong answers to be included.
	 * @return {object} Random question object from the list.
	 */
	const question = ({ wrong } = {}) => {
		if (!wrong) return QUESTIONS[randomNumber(AVAILABLE)];

		const _wrong = parameterLimitter(wrong, 'wrong', 1, 30);
		const _indexes = randomNumbers(_wrong, AVAILABLE);

		const _answers = [];

		for (let i = 0; i < _wrong; i += 1) {
			const _answer = QUESTIONS[_indexes?.[i]]?.answer;
			_answers.push(_answer || QUESTIONS[randomNumber(AVAILABLE)].answer);
		}

		return { ...QUESTIONS[randomNumber(AVAILABLE)], wrong: _answers };
	};

	/**
	 * Returns an array of random questions from the list.
	 *
	 * @param {object} [options] Guidelines for question generation.
	 * @param {number} [options.howMany=10] Number of question items to be generated.
	 * @param {number} [options.wrong] Number of wrong answers to be included.
	 * @return {array} Array of random question objects from the list.
	 */
	const questions = ({ howMany = 10, wrong } = {}) => {
		const _howMany = parameterLimitter(howMany, 'howMany', 1, 10000);
		const _wrong = wrong ? parameterLimitter(wrong, 'wrong', 1, 30) : undefined;

		const _result = [];

		for (let i = 0; i < _howMany; i += 1) {
			_result.push(question({ wrong: _wrong }));
		}

		return _result;
	};

	return { question, questions };
};
